### PR TITLE
fix(lib): Alternative Validation Approach

### PR DESCRIPTION
**Pull request by skorfmann**
While the `validation` hook seems to be a good approach to handle this, we should avoid using `toTerraform` to collect the required data. As an alternative this PR walks over all construct nodes and collects the data from the actual objects. At some point, we might wanna think about optimizing this a bit, so that we don't have to visit all nodes again and again - but for now I believe it's good enough.

For the error which gets thrown, we might want to add a handler which beautifies this a bit and gets rid of stacktraces, etc...

### CODE REVIEW

File: packages/cdktf/lib/terraform-stack.ts [hunk 1]
```diff
@@ -2,8 +2,6 @@ import { Construct, IConstruct, ISynthesisSession, Node } from "constructs";
 import { resolve } from "./_tokens";
 import * as fs from "fs";
 import * as path from "path";
-import camelcase = require("camelcase"); // ES module interop
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/terraform-stack.ts:5]: We can remove this from the bundledDependencies array in the package.json then

File: packages/cdktf/lib/validations/validate-provider-presence.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      return !this.foundProviders.some(
+        (p) => p.terraformGeneratorMetadata?.providerName === name
+      );
+    });
+
+    if (missingProviders.length === 0) {
+      return [];
+    } else {
+      return [
+        `Found resources without a matching povider. Please make sure to add the following providers to your stack ${missingProviders.join(
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/validations/validate-provider-presence.ts:44]:
```suggestion
        `Found resources without a matching povider. Please make sure to add the following providers to your stack: ${missingProviders.join(
```

**Review by DanielMSchmidt — APPROVED** [review]
Two small things, but seems like a great solution :)

File: packages/cdktf/lib/terraform-stack.ts (hunk 1 above)
Comment (skorfmann) [reply packages/cdktf/lib/terraform-stack.ts:5]: Good point - done 👍

File: packages/cdktf/lib/validations/validate-provider-presence.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      return !this.foundProviders.some(
+        (p) => p.terraformGeneratorMetadata?.providerName === name
+      );
+    });
+
+    if (missingProviders.length === 0) {
+      return [];
+    } else {
+      return [
+        `Found resources without a matching povider. Please make sure to add the following providers to your stack: ${missingProviders.join(
```
Comment (jsteinich) [inline packages/cdktf/lib/validations/validate-provider-presence.ts:44]: Perhaps in the future we could exclude providers that don't require any configuration.

Comment (DanielMSchmidt) [reply packages/cdktf/lib/validations/validate-provider-presence.ts:44]: I'm sure I ran into an issue where I forgot to initialized the docker provider and the program errored. The providers also help to set the right name in the required_providers (org is defaulted to hashicorp).

Comment (jsteinich) [reply packages/cdktf/lib/validations/validate-provider-presence.ts:44]: I think that is true, but I don't recall that being the case in hcl (maybe it changed in newer versions).

Comment (skorfmann) [reply packages/cdktf/lib/validations/validate-provider-presence.ts:44]: > Perhaps in the future we could exclude providers that don't require any configuration.

Do you know if that's documented somewhere?

Comment (jsteinich) [reply packages/cdktf/lib/validations/validate-provider-presence.ts:44]: Everything in the schema would be optional. Post Terraform 0.13, probably only works for providers in the `hashicorp` namespace.

**Comment by Ygshvr** [comment]
I am using cdktf 0.7.0 with python and facing a similar issue with the local provider.
I am trying to write to a file like this:
```python3
File(
            self,
            f"{cluster_name}-kubeconfig",
            content=self.gke_auth.kubeconfig_raw_output,
            filename=cluster_name,
            depends_on=[self.clusters[cluster_name]] # I have created an array of clusters so it's waiting for that cluster to be ready.
        )
```
But I get this error:
```
Found resources without a matching povider. Please make sure to add the following providers to your stack: local
```
So I tried to instantiate the provider like this 
```python3
LocalProvider(self, "local")
```
and got this error:
```
failed to instantiate provider "registry.terraform.io/hashicorp/local" to obtain schema: unknown provider "registry.terraform.io/hashicorp/local"
```
Any suggestions/workarounds?

**Comment by Ygshvr** [comment]
> @Ygshvr did you add the local provider to your `cdktf.json` file and run `cdktf get`?

Yes, I did.

**Comment by jsteinich** [comment]
@Ygshvr Can you please post your `cdktf.json` file? Also, are you behind any proxies?

**Comment by Ygshvr** [comment]
@jsteinich no, I am not behind a proxy. The code was downloaded successfully. I did check it by going into the imports directory. Here is the `cdktf.json`:
```json
{
  "language": "python",
  "app": "pipenv run python main.py",
  "terraformProviders": [
    "hashicorp/google@ ~> 4.0.0"
  ],
  "terraformModules": [
    "terraform-google-modules/kubernetes-engine/google//modules/auth"
  ],
  "codeMakerOutput": "imports",
  "projectId": "286db479-f6a5-4a90-88b4-b0f8903f245a"
}

```

**Comment by jsteinich** [comment]
@Ygshvr you'll either need to add the local provider to your `cdktf.json` and generate or use a pre-built version of it and then instantiate as you described earlier.

**Pull request merged**